File: demos/python-feequote/feequote/core/pricing.py
"""Core fee pricing.

Amounts are integer minor units (cents) and rates are integer basis points.
Nothing in this module touches binary floating point; the division is done
with integers so the result is exact for any amount the ledger can hold.
"""

from __future__ import annotations

from dataclasses import dataclass

BPS_PER_UNIT = 10_000


@dataclass(frozen=True)
class FeePolicy:
    """Bounds applied to a fee after it has been rounded to whole cents."""

    minimum_cents: int
    cap_cents: int

    def __post_init__(self) -> None:
        if self.minimum_cents < 0:
            raise ValueError("minimum_cents must not be negative")
        if self.cap_cents < self.minimum_cents:
            raise ValueError("cap_cents must not be below minimum_cents")

    def apply(self, fee_cents: int) -> int:
        """Clamp ``fee_cents`` into [minimum_cents, cap_cents]."""
        return max(self.minimum_cents, min(self.cap_cents, fee_cents))


DOMESTIC_RATE_BPS = 90
DOMESTIC_POLICY = FeePolicy(minimum_cents=100, cap_cents=2500)


def round_half_even_div(numerator: int, denominator: int) -> int:
    """Divide two non-negative integers and round the quotient half to even.

    This is banker's rounding on an exact rational, so 1111.5 becomes 1112,
    1112.5 becomes 1112 and 1111.05 becomes 1111.
    """
    if denominator <= 0:
        raise ValueError("denominator must be positive")
    if numerator < 0:
        raise ValueError("numerator must not be negative")
    quotient, remainder = divmod(numerator, denominator)
    doubled = remainder * 2
    if doubled > denominator or (doubled == denominator and quotient % 2 == 1):
        return quotient + 1
    return quotient


def price_with_policy(amount_cents: int, rate_bps: int, policy: FeePolicy) -> int:
    """Fee for ``amount_cents`` at ``rate_bps``, rounded, then bounded by ``policy``.

    This is the building block for every fee type. Rounding happens before
    the minimum and cap are applied, so the bounds always win.
    """
    if amount_cents < 0:
        raise ValueError("amount_cents must not be negative")
    if rate_bps < 0:
        raise ValueError("rate_bps must not be negative")
    raw_fee = round_half_even_div(amount_cents * rate_bps, BPS_PER_UNIT)
    return policy.apply(raw_fee)


def compute_fee(amount_cents: int) -> int:
    """Domestic transfer fee: 90 bps under the standard domestic policy."""
    return price_with_policy(amount_cents, DOMESTIC_RATE_BPS, DOMESTIC_POLICY)

File: demos/java-feequote/src/main/java/com/feequote/core/Pricing.java
package com.feequote.core;

/**
 * Core fee pricing.
 *
 * <p>Amounts are integer minor units (cents) and rates are integer basis points.
 * Nothing in this class touches binary floating point; the division is done with
 * integers so the result is exact for any amount the ledger can hold.
 */
public final class Pricing {

    /** Basis points in one whole unit, so 10000 bps is 100 percent. */
    public static final int BPS_PER_UNIT = 10_000;

    /** The domestic transfer rate, 90 basis points. */
    public static final int DOMESTIC_RATE_BPS = 90;

    /** Minimum 100 cents, cap 2500 cents. */
    public static final FeePolicy DOMESTIC_POLICY = new FeePolicy(100, 2500);

    private Pricing() {
    }

    /**
     * Divides two non-negative integers and rounds the quotient half to even.
     *
     * <p>This is banker's rounding on an exact rational, so 1111.5 becomes 1112,
     * 1112.5 becomes 1112 and 1111.05 becomes 1111.
     *
     * @param numerator the dividend, which must not be negative
     * @param denominator the divisor, which must be positive
     * @return the quotient rounded half to even
     */
    public static long roundHalfEvenDiv(long numerator, long denominator) {
        if (denominator <= 0) {
            throw new IllegalArgumentException("denominator must be positive");
        }
        if (numerator < 0) {
            throw new IllegalArgumentException("numerator must not be negative");
        }
        long quotient = numerator / denominator;
        long doubledRemainder = (numerator % denominator) * 2;
        if (doubledRemainder > denominator || (doubledRemainder == denominator && quotient % 2 == 1)) {
            return quotient + 1;
        }
        return quotient;
    }

    /**
     * Fee for {@code amountCents} at {@code rateBps}, rounded, then bounded by {@code policy}.
     *
     * <p>This is the building block for every fee type. Rounding happens before the
     * minimum and the cap are applied, so the bounds always win.
     *
     * @param amountCents the transfer amount in cents
     * @param rateBps the rate in basis points
     * @param policy the minimum and cap to apply after rounding
     * @return the fee in whole cents
     */
    public static long priceWithPolicy(long amountCents, int rateBps, FeePolicy policy) {
        if (amountCents < 0) {
            throw new IllegalArgumentException("amountCents must not be negative");
        }
        if (rateBps < 0) {
            throw new IllegalArgumentException("rateBps must not be negative");
        }
        long rawFee = roundHalfEvenDiv(Math.multiplyExact(amountCents, rateBps), BPS_PER_UNIT);
        return policy.apply(rawFee);
    }

    /** Domestic transfer fee: 90 bps under the standard domestic policy. */
    public static long computeFee(long amountCents) {
        return priceWithPolicy(amountCents, DOMESTIC_RATE_BPS, DOMESTIC_POLICY);
    }
}
